### PR TITLE
🌈 Toggle/Disable virtual console feature #1068

### PR DESCRIPTION
**Pull request by dance2die**
# 🌈 Feature

## What kind of change does this PR introduce?

This is a new feature to enable and disable (`toggles`) virtual console output.

## What is the current behavior?

There is no way to disable the CodeSandbox console output, which can cause the performance issue.
**Reference**: #1068

## What is the new behavior?

Adds a new `Preview` option, `Toggle console output`.

new option

## What steps did you take to test this?

I've only done manual test on the sandbox, but I'd love to know how I can test with `yarn test`.

manual test

________

### **Updated Video**: 

It shows "Console is enabled/disabled" messages (same format as "Console was cleared").

https://youtu.be/sIOxmNMPhZw

## Checklist

~- [ ] Documentation N/A - No preference doc is available~
    - I can do another PR with Preferences dialog documentation later.
- [x] Testing 
- [x] Ready to be merged 
- [x] Added myself to contributors table 

## Additional Context
1. "Toggle console output" and description thereof might not be appropriate. Would you advise how to word them?

2. Where would be a good place to add the documentation? (as I do not see [Docs](https://codesandbox.io/docs) regarding `Preferences` screen via Algolia search).

3. I might have to update the [contributing.md](https://github.com/codesandbox/codesandbox-client/blob/master/CONTRIBUTING.md) to add a workaround for the error `“NODE_ENV” is not recognized as an internal or external command, operable command or batch file` on Widnows 10.

My development platform is Windows 10.
winver

The error occurs while running `yarn test`, "initially".
yarn test error

To get around the issue, I've used [win-node-env](https://github.com/laggingreflex/win-node-env) as a workaround thus able to run `yarn test` successfully.
Refer to: [“NODE_ENV” is not recognized as an internal or external command, operable command or batch file](https://stackoverflow.com/a/40967643/4035)

yarn test sucess

### CODE REVIEW

File: packages/app/src/app/components/Preview/DevTools/Console/index.tsx [hunk 1]
```diff
@@ -284,11 +299,14 @@ const ConsoleFilterSelect = props => {
...
 };
 
+const ObservedConsole = inject('store', 'signals')(observer(Console));
+
 export default {
   id: 'codesandbox.console',
   title: 'Console',
   // @ts-ignore  TODO: fix this
-  Content: withTheme<StyledProps>(Console),
+  // Content: inject('store', 'signals')(observer(withTheme<StyledProps>(Console))),
```
Comment (dance2die) [inline packages/app/src/app/components/Preview/DevTools/Console/index.tsx]: Woops. I forgot to remove the old code. I will remote it soon.

**Comment by djD-REK** [comment]
Thanks @dance2die for the notes on the Windows deployment. Using Cmder and win-node-env together I was able to get a local deployment working. I was about to open a Linux VirtualBox, so it's good to know I don't need to (and can keep using GitHub Desktop, which is not available on Linux). I think you're right that it might be a good thing to add to the CONTRIBUTING.MD

**Comment by djD-REK** [comment]
> 
> 
> Thanks @dance2die for the notes on the Windows deployment. Using Cmder and win-node-env together I was able to get a local deployment working. I was about to open a Linux VirtualBox, so it's good to know I don't need to (and can keep using GitHub Desktop, which is not available on Linux). I think you're right that it might be a good thing to add to the CONTRIBUTING.MD

Actually mine still gives a crapton of warnings about case sensitivity because "c:" gets changed to "C:" and then when I go to launch the homepage I get this error:
image
```Failed to compile

GraphQL Error There was an error while compiling your site's GraphQL queries.
  Error: RelayParser: Encountered duplicate defintitions for one or more documents: each document must have a unique name. Duplicated documents:
- Docs
- Post
- Job
- cGitHubDjDRekCodesandboxClientPackagesHomepageSrcPagesBlogJs1226529959
- cGitHubDjDRekCodesandboxClientPackagesHomepageSrcPagesJobsJs238960520
- cGitHubDjDRekCodesandboxClientPackagesHomepageSrcPagesLegalPrivacyJs1983975423
- cGitHubDjDRekCodesandboxClientPackagesHomepageSrcPagesLegalTermsJs2629907227
- cGitHubDjDRekCodesandboxClientPackagesHomepageSrcPagesReactHooksJs2782515548
- cGitHubDjDRekCodesandboxClientPackagesHomepageSrcPagesWhoUsesCodesandboxJs2162632823
    

This error occurred during the build time and cannot be dismissed.
```
Any ideas?  I'm about to install Pop!OS in a VirtualBox and just develop on Linux for this project, I think it's going to be easier :+1:

**Comment by djD-REK** [comment]
> 
> 
> > Thanks @dance2die for the notes on the Windows deployment. Using Cmder and win-node-env together I was able to get a local deployment working. I was about to open a Linux VirtualBox, so it's good to know I don't need to (and can keep using GitHub Desktop, which is not available on Linux). I think you're right that it might be a good thing to add to the CONTRIBUTING.MD
> 
> Actually mine still gives a crapton of warnings about case sensitivity because "c:" gets changed to "C:" and then when I go to launch the homepage I get this error:
> image
> 
> ```
> 
> GraphQL Error There was an error while compiling your site's GraphQL queries.
>   Error: RelayParser: Encountered duplicate defintitions for one or more documents: each document must have a unique name. Duplicated documents:
> - Docs
> - Post
> - Job
> - cGitHubDjDRekCodesandboxClientPackagesHomepageSrcPagesBlogJs1226529959
> - cGitHubDjDRekCodesandboxClientPackagesHomepageSrcPagesJobsJs238960520
> - cGitHubDjDRekCodesandboxClientPackagesHomepageSrcPagesLegalPrivacyJs1983975423
> - cGitHubDjDRekCodesandboxClientPackagesHomepageSrcPagesLegalTermsJs2629907227
> - cGitHubDjDRekCodesandboxClientPackagesHomepageSrcPagesReactHooksJs2782515548
> - cGitHubDjDRekCodesandboxClientPackagesHomepageSrcPagesWhoUsesCodesandboxJs2162632823
>     
> 
> This error occurred during the build time and cannot be dismissed.
> ```
> 
> Any ideas? I'm about to install Pop!OS in a VirtualBox and just develop on Linux for this project, I think it's going to be easier +1

OK this happened because I had capital letters (uppercase) in my directory structure on Windows. Changing everything to lowercase fixed it. I'm actually sending a pull request over to gatsby-js to warn users if this error happens. Source: https://medium.com/r/?url=https%3A%2F%2Fgithub.com%2Fgatsbyjs%2Fgatsby%2Fissues%2F11688%23issuecomment-521933462

**Comment by dance2die** [comment]
Glad to see that your issue is resolved with building on Windows @djD-REK :)

**Comment by christianalfoni** [comment]
Ops, sorry, I accidentally clicked the close button 😂 We were talking about creating an async console, though this does not have a high priority at the moment. I think we should get this in @CompuIves? @dance2die Sorry this has taken a bit of time, you have time to update the PR? 😄

**Comment by dance2die** [comment]
Thank you for reviewing this PR @christianalfoni 😄 

I will update the PR with the latest master.

> We were talking about creating an async console

Sounds like a cool idea. I will keep up with issues/features~

**Comment by dance2die** [comment]
🤔 Let me check the errors and add patches.

I ran `yarn test` but I must've missed something.

image

**Comment by dance2die** [comment]
I will install `puppeteer` to figure out the integration test failure.

image

And as a side note, I will be adding a message `Console enabled/disabled` message similar to how `Console was cleared`.  
_I found not seeing a message was very confusing_.

image

**Comment by dance2die** [comment]
How could I monitor the "settings" change for the preference?

I was expecting the setting to cause `ConsoleComponent.UNSAFE_componentWillReceiveProps` to be triggered when I am injecting `store` with observer but the `cWRP` isn't triggered 😅 

```
inject('store')(observer(ConsoleComponent))
```

File: packages/app/src/app/components/Preview/DevTools/Console/index.tsx [hunk 2]
```diff
@@ -236,8 +275,12 @@ class ConsoleComponent extends React.Component<StyledProps> {
       searchKeywordsHasError = true;
     }
 
+    // We need to access  this.props.store.preferences.settings.toggleConsoleEnabled
+    // to trigger console enabled/disabled message
+    const isConsoleEnabled = extractToggleConsoleEnabled(this.props);
+
     return (
-      <Container>
+      <Container key={(isConsoleEnabled as unknown) as number}>
```
Comment (dance2die) [inline packages/app/src/app/components/Preview/DevTools/Console/index.tsx]: @christianalfoni Would this approach be correct to monitor `this.props.store.preferences.settings.toggleConsoleEnabled` value set in the global store to cause a re-render?
_Idea from https://overmindjs.org/guides/beginner/07_connectingcomponents?view=react&typescript=true#connecting-components-state-objects_

Is there a better way to "monitor" the settings value change?

**Comment by dance2die** [comment]
Probably not a good idea to merge this yet even though all the checks pass.

Would anyone help me where to look? 
Forking causes an issue after rebasing `master`

error

**Comment by dance2die** [comment]
I apologize for massive commits.
Is this commit list normal? (First time contributing to a big project like this).

____

The updated patch adds `Console is enabled/disabled` message.
Link to demo video (low quality): https://imgur.com/a/dXJy94x
High quality: https://youtu.be/sIOxmNMPhZw

**Comment by dance2die** [comment]
@CompuIves & @christianalfoni 

Am I doing something wrong as merging the upstream/master causes many commits and long time?  
Would you let me know if I should close this PR, and start a fresh PR?

**Comment by dance2die** [comment]
I decided to delete this PR and do another PR after the codebase is migrated to Overmind.  
And also this is old so keeping it up to date feels dirty, causing many commits (mostly due to my merge inability 😅).

**Comment by CompuIves** [comment]
Sorry for this! We'll make sure to merge the new PR asap.

**Comment by dance2die** [comment]
Thank you~
I will make it cleaner next time :)

**Pull request closed without merging**